### ISSUE TITLE
Race conditions during "install --prefer-source" when "git clone" times out

### ISSUE THREAD

**Issue by sustmi** [composer/composer#5800]
## What happens

When I run `composer install` I get the following output:

```
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
  - Installing doctrine/orm (dev-ddc1960-hotfix-and-ddc4005-hotfix ba2fa4b)
    Cloning ba2fa4bb174c467766ab6d0c205be7cf51ac395f
    Failed to download doctrine/orm from source: The process "git clone --no-checkout 'https://github.com/shopsys/doctrine2.git' '/home/me/project/vendor/doctrine/orm' && cd '/home/me/project/vendor/doctrine/orm' && git remote add composer 'https://github.com/shopsys/doctrine2.git' && git fetch composer" exceeded the timeout of 300 seconds.
    Now trying to download from dist
  - Installing doctrine/orm (dev-ddc1960-hotfix-and-ddc4005-hotfix ba2fa4b)
    Loading from cache
...
```

Composer finishes successfully and everything looks great.

The problem is that after approximately 8 minutes directory `vendor/doctrine/orm` disappears.
## Why it happens

I figured out that the cause of this problem is that `git clone` of `doctrine/orm` is very slow today (it takes around 13 minutes).
After 5 minutes `GitDownloader` times out and `Process::stop()` is called.
The process tree before stopping `GitDownloader` looks like this:

```
php composer install
   \_ sh -c git clone --no-checkout 'https://github.com/shopsys/doctrine2.git' '/home/me/project/vendor/doctrine/orm' && cd '/home/me/project/vendor/doctrine/orm' && git remote add composer 'https://github.com/shopsys/doctrine2.git' && git fetch composer
       \_ git clone --no-checkout https://github.com/shopsys/doctrine2.git /home/me/project/vendor/doctrine/orm
          \_ git-remote-https origin https://github.com/shopsys/doctrine2.git
               \_ git fetch-pack --stateless-rpc --stdin --lock-pack --thin --check-self-contained-and-connected --cloning --no-progress https://github.com/shopsys/doctrine2.git/
```

After stopping `GitDownloader`, the process tree looks like this:

```
php composer install
    \_ <composer installing doctrine/orm from dist>
    \_ git clone --no-checkout https://github.com/shopsys/doctrine2.git /home/me/project/vendor/doctrine/orm
        \_ git-remote-https origin https://github.com/shopsys/doctrine2.git
            \_ git fetch-pack --stateless-rpc --stdin --lock-pack --thin --check-self-contained-and-connected --cloning --no-progress https://github.com/shopsys/doctrine2.git/
```

As you can see the `sh -c` process was killed but its children are still running. This is normal behaviour in Linux (see http://riccomini.name/posts/linux/2012-09-25-kill-subprocesses-linux-bash/ )

Composer then tries to download `doctrine/orm` from `dist`.
That means that `FileDownloader::doDownload()` is called. The very first line of this method is:

``` php
$this->filesystem->emptyDirectory($path);
```

This effectively deletes the `vendor/doctrine/orm`.
Note that `git clone` is still running in background working with `vendor/doctrine/orm/.git` directory.

Then Composer successfully finishes and the process tree looks like this:

```
git clone --no-checkout https://github.com/shopsys/doctrine2.git /home/me/project/vendor/doctrine/orm
    \_ git-remote-https origin https://github.com/shopsys/doctrine2.git
        \_ git fetch-pack --stateless-rpc --stdin --lock-pack --thin --check-self-contained-and-connected --cloning --no-progress https://github.com/shopsys/doctrine2.git/
```

The `vendor/doctrine/orm` directory is all right at this time. But after approximately 8 minutes `git clone` finishes downloading from Github and fails because `vendor/doctrine/orm/.git` directory was deleted meanwhile.
And when `git clone` fails it deletes the the `.git` directory and working directory (see https://github.com/git/git/blob/master/builtin/clone.c#L935-L936 and https://github.com/git/git/blob/master/builtin/clone.c#L479-L488 ).

That means that `vendor/doctrine/orm` is deleted again and whole application is broken because of missing source files.
---
# EDIT (workaround and test-case)
## Workaround

Add this to your `composer.json` to prefer not running `git clone` when `dist` source is available:

```
{
    "config": {
        "preferred-install": "dist"
    }
}
```

Or use:

``` bash
composer install --prefer-dist
```
### 1. Create composer.json:

``` json
{
    "require": {
        "doctrine/orm": "dev-master"
    },
    "config": {
        "process-timeout": 5
    }
}
```
### 2. Run:

``` bash
$ composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
  - Installing doctrine/orm (dev-master d6eddab)
    Cloning d6eddab94fd71f18ee80b9d7f2a26fd4b32b246c
    Failed to download doctrine/orm from source: The process "git clone --no-checkout 'https://github.com/doctrine/doctrine2.git' '/home/me/project/vendor/doctrine/orm' && cd '/home/me/project/vendor/doctrine/orm' && git remote add composer 'https://github.com/doctrine/doctrine2.git' && git fetch composer" exceeded the timeout of 5 seconds.
    Now trying to download from dist
  - Installing doctrine/orm (dev-master d6eddab)
    Loading from cache

    Skipped installation of bin bin/doctrine for package doctrine/orm: name conflicts with an existing file
doctrine/orm suggests installing symfony/yaml (If you want to use YAML Metadata Mapping Driver)
Generating autoload files
```
### 3. Quickly check that `git clone` is still running:

``` bash
$ ps -ft
  PID TTY      STAT   TIME COMMAND
13421 pts/4    Ss     0:00 bash
14284 pts/4    R+     0:00  \_ ps -ft
14268 pts/4    S      0:00 git clone --no-checkout https://github.com/doctrine/doctrine2.git /home/me/project/vendor/doctrine/orm
14269 pts/4    S      0:00  \_ git-remote-https origin https://github.com/doctrine/doctrine2.git
14271 pts/4    Sl     0:00      \_ git fetch-pack --stateless-rpc --stdin --lock-pack --thin --check-self-contained-and-connected --cloning --no-progress https://github.com/doctrine/doctrine2.git/
14273 pts/4    S      0:00          \_ git index-pack --stdin --fix-thin --keep=fetch-pack 14271 on localhost.localdomain --check-self-contained-and-connected --pack_header=2,113053
14261 pts/4    S      0:00 git-remote-https origin https://github.com/doctrine/doctrine2.git
14263 pts/4    Sl     0:00  \_ git fetch-pack --stateless-rpc --stdin --lock-pack --thin --check-self-contained-and-connected --cloning --no-progress https://github.com/doctrine/doctrine2.git/
14265 pts/4    S      0:00      \_ git index-pack --stdin --fix-thin --keep=fetch-pack 14263 on localhost.localdomain --check-self-contained-and-connected --pack_header=2,138024
```
### 4. Wait for `git clone` to finish and see that `vendor/doctrine/orm` directory does not exists:

```
$ ps -ft
  PID TTY      STAT   TIME COMMAND
13421 pts/4    Ss     0:00 bash
14380 pts/4    R+     0:00  \_ ps -ft

$ ls -la vendor/doctrine/orm
ls: cannot access 'vendor/doctrine/orm': No such file or directory
```
## Summary of possible ways to fix this problem:
- Make `Composer` not to run `git clone` as part of compound command
- Make `Symfony/Process` component kill child processes on `stop()`
- Make PHP [kill not only wrapper but the real process too on `proc_terminate()`](https://bugs.php.net/bug.php?id=33505)
- `git clone` into temporary directory so that [GITs junk removal](https://github.com/git/git/blob/master/builtin/clone.c#L935) does not delete files installed from `dist`

**Comment by sustmi** [comment]
This issue can be related with https://github.com/composer/composer/issues/5798 .

**Comment by sustmi** [comment]
As a workaround `composer install --prefer-dist` can be used because it usually ends up in not using `git clone`.

**Comment by stof** [comment]
the fact that stopping a process kills only the sh wrapper and not the actual command being run looks likea bug in symfony/process IMO. Can you open an issue on http://github.com/symfony/symfony about this ?

**Comment by sustmi** [comment]
@stof: Do you think that `Process::stop()` should kill all the child processes?
Because if `git clone` did not delete the directories in its main process but for example in `fetch-pack` sub-process then killing `git clone` instead `sh` wrapper would not help (because `fetch-pack` will be still running).

**Comment by sustmi** [comment]
The default behaviour when killing processes in Linux is to leave the child processes running.
Only `bash` does some magic that when you press `CTRL+C` even child processes are killed (see http://stackoverflow.com/questions/6108953/how-does-ctrl-c-terminate-a-child-process ).

**Comment by sustmi** [comment]
I guess that it is related to this PHP bug: https://bugs.php.net/bug.php?id=39992

**Comment by sustmi** [comment]
The problem with killing `sh` wrapper instead of real process in `Process` component was discussed here: https://github.com/symfony/symfony/issues/5499

**Comment by alcohol** [comment]
Not much we can do about this, expect for mentioning that you can raise the timeout:
https://getcomposer.org/doc/03-cli.md#composer-process-timeout
https://getcomposer.org/doc/06-config.md#process-timeout

**Comment by staabm** [comment]
Seems like there is a documenzed workaround
https://github.com/symfony/symfony-docs/pull/2504#discussion-diff-3786116R192

**Comment by sustmi** [comment]
@staabm: Unfortunately, this workaround does not work for compound statements like: 
```bash
git clone --no-checkout 'https://github.com/shopsys/doctrine2.git' '/home/me/project/vendor/doctrine/orm' && cd '/home/me/project/vendor/doctrine/orm' && git remote add composer 'https://github.com/shopsys/doctrine2.git' && git fetch composer
```
that Composer uses.

Try this:
```bash
# creates "test" directory
echo a && mkdir test

# does not create "test" directory (mkdir is "swallowed")
exec echo a && mkdir test
```

**Comment by alcohol** [comment]
Correct, you would run that as

``` sh
exec bash -c 'echo a && mkdir test'
```

**Comment by sustmi** [comment]
@alcohol: But then we are back to the original problem. `bash` wraps the inner process and when you call `proc_terminate()`, the children processes will be still running.
You try this:
```php
<?php

$cmd = 'exec bash -c \'sleep 10 && mkdir "test"\'';
$proc = proc_open($cmd, [['pipe', 'r'], ['pipe', 'w'], ['pipe', 'w']], $pipes);

proc_terminate($proc);
```

```bash
$ ps -ft
  PID TTY      STAT   TIME COMMAND
 5269 pts/2    Ss     0:00 bash
 6160 pts/2    R+     0:00  \_ ps -ft
 6152 pts/2    S      0:00 sleep 10
```

**Comment by sustmi** [comment]
The problem goes back to year 2005 and it is not fixed yet: https://bugs.php.net/bug.php?id=33505 .
To fix this issue i suggest:
* not use compound commands in Composer because simple statements are not executed in `sh` wrapper (I guess that `sh` calls `exec()` instead of `fork()` when it detects that it is simple statement)
* implement something like `stopTree()` in `Process` component to enable killing process including child processes

**Comment by alcohol** [comment]
Do not really see the need for this, since it just adds complexity for an edge case that could just as easily be mitigated by extending the process timeout value.

**Comment by sustmi** [comment]
@alcohol: It happened for the first time to us but we spend many hours debbuging the issue.
It also broke out CI server.
We use default Composer settings (no `--prefer-dist`) on our production server and if we did deploy yesterday.
We use:
```json
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/shopsys/doctrine2.git"
        },
    ],
    "require": {
        "doctrine/orm": "dev-ddc1960-hotfix-and-ddc4005-hotfix as 2.5.1"
    }
}
```
and this probably makes Composer try to download from `source` by default.

Shouldn't we at least document that such combination can cause broken `vendor` directory?
Because I think that it is fatal error if Composer finishes successfully but `vendor` directory is broken.

Do you really think this is an edge case?

**Comment by sustmi** [comment]
Making `dist` the default even for non-stable packages and documenting the possible issues when using `source` would be sufficient to make me stay in belief that Composer is production ready.

**Comment by alcohol** [comment]
In my personal experience and opinion, using Composer on production is a huge mistake. Composer is not a deployment tool. It is a development/build tool. Anyone using it on production should do so at their own risk completely.

I do not see us making `dist` the default for non-stable releases. We default to `source` for those because it makes updating them more convenient/efficient (they change too frequently to download a new archive every time you run update). Using non-stable releases on production should already fail any sanity-checks.

Also, `--prefer-dist` **is the default**.

**Comment by sustmi** [comment]
@alcohol: Thank you for your insight about not using Composer in production.

Since which version is `--prefer-dist` option default? I have `Composer version 1.2.1 2016-09-12 11:27:19` and `composer install` downloads non stable packages from sources while `composer install --prefer-dist` does not.
Also [documentation](https://getcomposer.org/doc/03-cli.md#install) says:
>--prefer-source: There are two ways of downloading a package: source and dist. **For stable versions Composer will use the dist by default**. The source is a version control repository. If --prefer-source is enabled, Composer will install from source if there is one. This is useful if you want to make a bugfix to a project and get a local git clone of the dependency directly.

So I guess that non stable packages are primarily installed from sources.
See also: https://getcomposer.org/doc/06-config.md#preferred-install . It says that `auto` is default which (I guess; not documented) does what I describe.

**Comment by stof** [comment]
@alcohol it is *not* the default.
However, you can make it the default on your system through configuration: https://getcomposer.org/doc/06-config.md#preferred-install

**Comment by alcohol** [comment]
My bad. I almost never use non-stable packages, so the only times I got packages installed from source was when I explicitly used `--prefer-source`. Therefor I just assumed dist is the default. Kind of hard to follow the myriad of options and what that explicitly mean, since so many got piled up over time.

**Comment by sustmi** [comment]
Sorry for being so insisting. I did not meant to be rude.
I "fixed" the issue in our environment. Using `"preferred-install": "dist"` in our `composer.json`s.
I just wanted to help the others so they will not run into the same debugging nightmare.
Lets hope that no more users will run into this peculiar issue.

**Comment by alcohol** [comment]
Rude? Not at all. I can understand why this is not exactly desirable behavior. Unfortunately we simply cannot provide the perfect workflow for each and every edge case. We have to find a sensible middle ground.

**Comment by sustmi** [comment]
I updated the issue with the best available workaround and test case.

**Comment by alcohol** [comment]
> git clone into temporary directory so that GITs junk removal does not delete files installed from dist

This would be my preferred solution actually. But since we chain/compound commands, not sure how feasible it will be in the end.

**Comment by Seldaek** [comment]
Closing as I don't really see a viable way to improve this, given it's the first time that someone reports it I doubt it's a widespread enough issue that it would justify the added complexity. It's good to have the issue summing up everything though in case anyone stumbles upon it in the future.
